File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/ObjectMemberInitializerSyntax.cs
using System;
using Microsoft.CodeAnalysis.Syntax.InternalSyntax;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class ObjectMemberInitializerSyntax : ObjectCreationInitializerSyntax
	{
		internal readonly KeywordSyntax _withKeyword;

		internal readonly PunctuationSyntax _openBraceToken;

		internal readonly GreenNode _initializers;

		internal readonly PunctuationSyntax _closeBraceToken;

		internal static Func<ObjectReader, object> CreateInstance;

		internal KeywordSyntax WithKeyword => _withKeyword;

		internal PunctuationSyntax OpenBraceToken => _openBraceToken;

		internal Microsoft.CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList<FieldInitializerSyntax> Initializers => new Microsoft.CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList<FieldInitializerSyntax>(new Microsoft.CodeAnalysis.Syntax.InternalSyntax.SyntaxList<FieldInitializerSyntax>(_initializers));

		internal PunctuationSyntax CloseBraceToken => _closeBraceToken;

		internal ObjectMemberInitializerSyntax(SyntaxKind kind, KeywordSyntax withKeyword, PunctuationSyntax openBraceToken, GreenNode initializers, PunctuationSyntax closeBraceToken)
			: base(kind)
		{
			base._slotCount = 4;
			AdjustFlagsAndWidth(withKeyword);
			_withKeyword = withKeyword;
			AdjustFlagsAndWidth(openBraceToken);
			_openBraceToken = openBraceToken;
			if (initializers != null)
			{
				AdjustFlagsAndWidth(initializers);
				_initializers = initializers;
			}
			AdjustFlagsAndWidth(closeBraceToken);
			_closeBraceToken = closeBraceToken;
		}

		internal ObjectMemberInitializerSyntax(SyntaxKind kind, KeywordSyntax withKeyword, PunctuationSyntax openBraceToken, GreenNode initializers, PunctuationSyntax closeBraceToken, ISyntaxFactoryContext context)
			: base(kind)
		{
			base._slotCount = 4;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(withKeyword);
			_withKeyword = withKeyword;
			AdjustFlagsAndWidth(openBraceToken);
			_openBraceToken = openBraceToken;
			if (initializers != null)
			{
				AdjustFlagsAndWidth(initializers);
				_initializers = initializers;
			}
			AdjustFlagsAndWidth(closeBraceToken);
			_closeBraceToken = closeBraceToken;
		}

		internal ObjectMemberInitializerSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax withKeyword, PunctuationSyntax openBraceToken, GreenNode initializers, PunctuationSyntax closeBraceToken)
			: base(kind, errors, annotations)
		{
			base._slotCount = 4;
			AdjustFlagsAndWidth(withKeyword);
			_withKeyword = withKeyword;
			AdjustFlagsAndWidth(openBraceToken);
			_openBraceToken = openBraceToken;
			if (initializers != null)
			{
				AdjustFlagsAndWidth(initializers);
				_initializers = initializers;
			}
			AdjustFlagsAndWidth(closeBraceToken);
			_closeBraceToken = closeBraceToken;
		}

		internal ObjectMemberInitializerSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 4;
			KeywordSyntax keywordSyntax = (KeywordSyntax)reader.ReadValue();
			if (keywordSyntax != null)
			{
				AdjustFlagsAndWidth(keywordSyntax);
				_withKeyword = keywordSyntax;
			}
			PunctuationSyntax punctuationSyntax = (PunctuationSyntax)reader.ReadValue();
			if (punctuationSyntax != null)
			{
				AdjustFlagsAndWidth(punctuationSyntax);
				_openBraceToken = punctuationSyntax;
			}
			GreenNode greenNode = (GreenNode)reader.ReadValue();
			if (greenNode != null)
			{
				AdjustFlagsAndWidth(greenNode);
				_initializers = greenNode;
			}
			PunctuationSyntax punctuationSyntax2 = (PunctuationSyntax)reader.ReadValue();
			if (punctuationSyntax2 != null)
			{
				AdjustFlagsAndWidth(punctuationSyntax2);
				_closeBraceToken = punctuationSyntax2;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_withKeyword);
			writer.WriteValue(_openBraceToken);
			writer.WriteValue(_initializers);
			writer.WriteValue(_closeBraceToken);
		}

		static ObjectMemberInitializerSyntax()
		{
			CreateInstance = (ObjectReader o) => new ObjectMemberInitializerSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(ObjectMemberInitializerSyntax), (ObjectReader r) => new ObjectMemberInitializerSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectMemberInitializerSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _withKeyword, 
				1 => _openBraceToken, 
				2 => _initializers, 
				3 => _closeBraceToken, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new ObjectMemberInitializerSyntax(base.Kind, newErrors, GetAnnotations(), _withKeyword, _openBraceToken, _initializers, _closeBraceToken);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new ObjectMemberInitializerSyntax(base.Kind, GetDiagnostics(), annotations, _withKeyword, _openBraceToken, _initializers, _closeBraceToken);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitObjectMemberInitializer(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/ObjectMemberInitializerSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Syntax.InternalSyntax
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class ObjectMemberInitializerSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectCreationInitializerSyntax
		Friend ReadOnly _withKeyword As KeywordSyntax

		Friend ReadOnly _openBraceToken As PunctuationSyntax

		Friend ReadOnly _initializers As GreenNode

		Friend ReadOnly _closeBraceToken As PunctuationSyntax

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property CloseBraceToken As PunctuationSyntax
			Get
				Return Me._closeBraceToken
			End Get
		End Property

		Friend ReadOnly Property Initializers As Microsoft.CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.FieldInitializerSyntax)
			Get
				Return New Microsoft.CodeAnalysis.Syntax.InternalSyntax.SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.FieldInitializerSyntax)(New Microsoft.CodeAnalysis.Syntax.InternalSyntax.SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.FieldInitializerSyntax)(Me._initializers))
			End Get
		End Property

		Friend ReadOnly Property OpenBraceToken As PunctuationSyntax
			Get
				Return Me._openBraceToken
			End Get
		End Property

		Friend ReadOnly Property WithKeyword As KeywordSyntax
			Get
				Return Me._withKeyword
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectMemberInitializerSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectMemberInitializerSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectMemberInitializerSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectMemberInitializerSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal withKeyword As KeywordSyntax, ByVal openBraceToken As PunctuationSyntax, ByVal initializers As GreenNode, ByVal closeBraceToken As PunctuationSyntax)
			MyBase.New(kind)
			MyBase._slotCount = 4
			MyBase.AdjustFlagsAndWidth(withKeyword)
			Me._withKeyword = withKeyword
			MyBase.AdjustFlagsAndWidth(openBraceToken)
			Me._openBraceToken = openBraceToken
			If (initializers IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(initializers)
				Me._initializers = initializers
			End If
			MyBase.AdjustFlagsAndWidth(closeBraceToken)
			Me._closeBraceToken = closeBraceToken
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal withKeyword As KeywordSyntax, ByVal openBraceToken As PunctuationSyntax, ByVal initializers As GreenNode, ByVal closeBraceToken As PunctuationSyntax, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind)
			MyBase._slotCount = 4
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(withKeyword)
			Me._withKeyword = withKeyword
			MyBase.AdjustFlagsAndWidth(openBraceToken)
			Me._openBraceToken = openBraceToken
			If (initializers IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(initializers)
				Me._initializers = initializers
			End If
			MyBase.AdjustFlagsAndWidth(closeBraceToken)
			Me._closeBraceToken = closeBraceToken
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal withKeyword As KeywordSyntax, ByVal openBraceToken As PunctuationSyntax, ByVal initializers As GreenNode, ByVal closeBraceToken As PunctuationSyntax)
			MyBase.New(kind, errors, annotations)
			MyBase._slotCount = 4
			MyBase.AdjustFlagsAndWidth(withKeyword)
			Me._withKeyword = withKeyword
			MyBase.AdjustFlagsAndWidth(openBraceToken)
			Me._openBraceToken = openBraceToken
			If (initializers IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(initializers)
				Me._initializers = initializers
			End If
			MyBase.AdjustFlagsAndWidth(closeBraceToken)
			Me._closeBraceToken = closeBraceToken
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 4
			Dim keywordSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax)
			If (keywordSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(keywordSyntax)
				Me._withKeyword = keywordSyntax
			End If
			Dim punctuationSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			If (punctuationSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(punctuationSyntax)
				Me._openBraceToken = punctuationSyntax
			End If
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.GreenNode)
			If (greenNode IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(greenNode)
				Me._initializers = greenNode
			End If
			Dim punctuationSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			If (punctuationSyntax1 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(punctuationSyntax1)
				Me._closeBraceToken = punctuationSyntax1
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitObjectMemberInitializer(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.ObjectMemberInitializerSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Select Case i
				Case 0
					greenNode = Me._withKeyword
					Exit Select
				Case 1
					greenNode = Me._openBraceToken
					Exit Select
				Case 2
					greenNode = Me._initializers
					Exit Select
				Case 3
					greenNode = Me._closeBraceToken
					Exit Select
				Case Else
					greenNode = Nothing
					Exit Select
			End Select
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectMemberInitializerSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._withKeyword, Me._openBraceToken, Me._initializers, Me._closeBraceToken)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ObjectMemberInitializerSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._withKeyword, Me._openBraceToken, Me._initializers, Me._closeBraceToken)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._withKeyword, IObjectWritable))
			writer.WriteValue(DirectCast(Me._openBraceToken, IObjectWritable))
			writer.WriteValue(DirectCast(Me._initializers, IObjectWritable))
			writer.WriteValue(DirectCast(Me._closeBraceToken, IObjectWritable))
		End Sub
	End Class
End Namespace